library/src/layout/par.rs

```diff
@@ -1137,12 +1149,20 @@ fn line<'a>(
         // need the shaped empty string to make the line the appropriate
         // height. That is the case exactly if the string is empty and there
         // are no other items in the line.
-        if hyphen || start + shaped.text.len() > range.end {
+        if hyphen || start + shaped.text.len() > range.end || end_cjk_punct {
             if hyphen || start < range.end || before.is_empty() {
                 let mut reshaped = shaped.reshape(vt, &p.spans, start..range.end);
                 if hyphen || shy {
                     reshaped.push_hyphen(vt);
                 }
+                let punct = reshaped.glyphs.to_mut().last_mut();
```

Comment:
this `to_mut` is potentially expensive. we want to reuse results from the initial shaping pass as much as possible. we should only actually do this if necessary. (so maybe check the condition first)

Reply:
Fixed in 9f3653fe45fc715dc1843ba12b6f5b28ada9146d